django/db/models/indexes.py

```diff
@@ -60,10 +75,91 @@ def _get_condition_sql(self, model, schema_editor):
+            expression = expression.get_source_expressions()[0]
+        expression_sql, params = compiler.compile(expression)
+        params = tuple(map(schema_editor.quote_value, params))
+        # Wrap expression in parentheses
+        expression_sql = '(%s)' % expression_sql
+        return expression_sql % params, suffix
+
+    def _get_ordered_field(self, model, field_name):
+        """
+        Return the field for supplied name and an optional ordering
+        """
+        field_name, order = (
+            (field_name[1:], 'DESC') if field_name.startswith('-') else (field_name, '')
+        )
+        field = model._meta.get_field(field_name)
+        return field, order
+
+    def _get_field_orders(self, model, schema_editor, using):
+        """
+        Return fields and expressions together with their ordering
+        """
+        fields = []
+        for field in self.fields:
+            if field in self.expressions:
+                field, order = self._get_ordered_expression(
+                    model, schema_editor, using, field
+                )
+            else:
+                field, order = self._get_ordered_field(model, field)
+            fields.append((field, order))
+        return fields
+
+    def _validate_supports_expression_indexes(self, schema_editor):
+        """
+        Validate that database supports supplied expressions
+        """
+        connection = schema_editor.connection
+        supports_expression_indexes = connection.features.supports_expression_indexes
+        for column_expression in self.expressions:
+            if (
+                not supports_expression_indexes and
+                hasattr(column_expression, 'flatten') and
+                any(
+                    isinstance(expr, (BaseExpression, F))
+                    for expr in column_expression.flatten()
+                )
+            ):
+                raise NotSupportedError(
+                    (
+                        'Not creating expression index:\n'
+                        '   {expression}\n'
+                        'Expression indexes are not supported on {vendor}.'
+                    ).format(
+                        expression=column_expression, vendor=connection.display_name
+                    )
+                )
+
+    def create_sql(self, model, schema_editor, using="", **kwargs):
+        try:
+            self._validate_supports_expression_indexes(schema_editor)
+        except NotSupportedError as e:
+            # While it's an error on platforms w/o support for expression
+            # indexes to create one, the presence of an index is often not
+            # required. Thus, we're raising a warning instead of blowing up.
+            # That also seems to be the only way to allow 3rd party packages
+            # using expression indexes while supporting all databases for
+            # everything else.
+            warnings.warn(str(e), RuntimeWarning)
+            return None
```

Comment:
I think we should move this and `_validate_supports_expression_indexes()` to a system check `_check_indexes()` with `hint` that _"An index won't be created. Silence this warning if you don't care about it."_ (like we do for constraints).